File: dotnet-desktop-guide/framework/winforms/controls/snippets/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control/cs/Form1.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Windows.Forms;

namespace project
{
    public partial class Form1 : Form
    {
        //List of all students
        private ArrayList studentArray = new ArrayList();

        //Windows Form application
        public Form1()
        {
            InitializeComponent();

            //Loading the TreeView with data
            PopulateTreeView();

            //Below method implementations will start printing node values to MessageBoxes and can be time consuming.
            //An alternate way to check is adding two labels and printing the values of the total nodes in each using both the below traversal methods.
            
            //Printing the total number of nodes recursively
            //CallRecursive(trvTest);

            //CallNonRecursive(trvTest);
        }

        // <PrintRecursive>
        private void PrintRecursive(TreeNode treeNode)
        {
            // Print the node.  
            System.Diagnostics.Debug.WriteLine(treeNode.Text);
            MessageBox.Show(treeNode.Text);

            // Visit each node recursively.  
            foreach (TreeNode tn in treeNode.Nodes)
            {
                PrintRecursive(tn);
            }
        }       

        // Call the procedure using the TreeView.  
        private void CallRecursive(TreeView treeView)
        {
            // Print each node recursively.  
            foreach (TreeNode n in treeView.Nodes)
            {
                //recursiveTotalNodes++;
                PrintRecursive(n);
            }
        }
        // </PrintRecursive>

        // <PrintNonRecursive>
        private void PrintNonRecursive(TreeNode treeNode)
        {
            if (treeNode != null)
            {
                //Using a queue to store and process each node in the TreeView
                Queue<TreeNode> staging = new Queue<TreeNode>();
                staging.Enqueue(treeNode);

                while (staging.Count > 0)
                {
                    treeNode = staging.Dequeue();
                    
                    // Print the node.  
                    System.Diagnostics.Debug.WriteLine(treeNode.Text);
                    MessageBox.Show(treeNode.Text);

                    foreach (TreeNode node in treeNode.Nodes)
                    {
                        staging.Enqueue(node);
                    }
                }
            }
        }

        // Call the procedure using the TreeView.  
        private void CallNonRecursive(TreeView treeView)
        {
            // Print each node.
            foreach (TreeNode n in treeView.Nodes)
            {
                PrintNonRecursive(n);
            }
        }
        // </PrintNonRecursive>

        private void PopulateTreeView()
        {
            // Add students  to the ArrayList of Student objects.
            for (int x = 0; x < 5; x++)
            {
                studentArray.Add(new Student("Student " + x.ToString()));
            }

            // Add subjects to each Student object in the ArrayList.
            foreach (Student student in studentArray)
            {
                for (int y = 0; y < 10; y++)
                {
                    student.StudentSubjects.Add(new Subject("Subject " + y.ToString()));
                }
            }

            Random _random = new Random();

            // Add subjects to each Student object in the ArrayList.
            foreach (Student student in studentArray)
            {
                foreach (Subject subject in student.StudentSubjects)
                {
                    var gen = _random.Next(1, 10);

                    for (int y = 0; y < gen; y++)
                    {
                        subject.SubjectTextbooks.Add(new TextBook("TextBook " + y.ToString()));
                    }
                }                
            }

            // Suppress repainting the TreeView until all the objects have been created.
            trvTest.BeginUpdate();

            // Clear the TreeView each time the method is called.
            trvTest.Nodes.Clear();

            // Add a root TreeNode for each Student object in the ArrayList.
            foreach (Student student in studentArray)
            {
                trvTest.Nodes.Add(new TreeNode(student.StudentName));

                // Add a child treenode for each Subject object in the current student object.
                foreach (Subject subject in student.StudentSubjects)
                {
                    trvTest.Nodes[studentArray.IndexOf(student)].Nodes.Add(
                      new TreeNode(subject.SubjectID));

                    // Add a child treenode for each TextBook object in the current Subject object.

                    foreach (TextBook textbook in subject.SubjectTextbooks)
                    {
                        trvTest.Nodes[studentArray.IndexOf(student)].Nodes[student.StudentSubjects.IndexOf(subject)].Nodes.Add(
                        new TreeNode(textbook.TextBookID));
                    }
                }
            }

            // Begin repainting the TreeView.
            trvTest.EndUpdate();
        }


    }
}

File: dotnet-desktop-guide/framework/winforms/controls/snippets/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control/vb/Form1.vb
Imports project.project

Public Class Form1
    Private studentArray As ArrayList = New ArrayList()

    Public Sub New()
        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.
        'Loading the TreeView with data
        PopulateTreeView()

        'Below method implementations will start printing node values to MessageBoxes and can be time consuming.
        'An alternate way to check is adding two labels and printing the values of the total nodes in each using both the below traversal methods.
         
        'Print all the nodes recursively
        'CallRecursive(trvTest)
        'CallNonRecursive(trvTest)
    End Sub

    '<PrintRecursive>
    Private Sub PrintRecursive(n As TreeNode)
        System.Diagnostics.Debug.WriteLine(n.Text)
        MessageBox.Show(n.Text)
        Dim aNode As TreeNode
        For Each aNode In n.Nodes
            PrintRecursive(aNode)
        Next
    End Sub

    ' Call the procedure using the top nodes of the treeview.  
    Private Sub CallRecursive(aTreeView As TreeView)
        Dim n As TreeNode
        For Each n In aTreeView.Nodes
            PrintRecursive(n)
        Next
    End Sub
    '</PrintRecursive>

    '<PrintNonRecursive>
    Private Sub PrintNonrecursive(n As TreeNode)
        If n IsNot Nothing Then
            Dim staging As Queue(Of TreeNode) = New Queue(Of TreeNode)
            staging.Enqueue(n)
            While staging.Count > 0
                n = staging.Dequeue()

                'Print the node.  
                System.Diagnostics.Debug.WriteLine(n.Text)
                MessageBox.Show(n.Text)

                Dim node As TreeNode
                For Each node In n.Nodes
                    staging.Enqueue(node)
                Next
            End While
        End If
    End Sub

    Private Sub CallNonRecursive(aTreeView As TreeView)
        Dim n As TreeNode
        For Each n In aTreeView.Nodes
            PrintNonrecursive(n)
        Next
    End Sub
    '</PrintNonRecursive>

    Private Sub PopulateTreeView()
        For x As Integer = 0 To 4
            studentArray.Add(New Student("Student " & x.ToString()))
        Next

        For Each student As Student In studentArray

            For y As Integer = 0 To 9
                student.StudentSubjects.Add(New Subject("Subject " & y.ToString()))
            Next
        Next

        Dim _random As Random = New Random()

        For Each student As Student In studentArray

            For Each subject As Subject In student.StudentSubjects
                Dim gen = _random.[Next](1, 10)

                For y As Integer = 0 To gen - 1
                    subject.SubjectTextbooks.Add(New TextBook("TextBook " & y.ToString()))
                Next
            Next
        Next

        trvTest.BeginUpdate()
        trvTest.Nodes.Clear()

        For Each student As Student In studentArray
            trvTest.Nodes.Add(New TreeNode(student.StudentName))

            For Each subject As Subject In student.StudentSubjects
                trvTest.Nodes(studentArray.IndexOf(student)).Nodes.Add(New TreeNode(subject.SubjectID))

                For Each textbook As TextBook In subject.SubjectTextbooks
                    trvTest.Nodes(studentArray.IndexOf(student)).Nodes(student.StudentSubjects.IndexOf(subject)).Nodes.Add(New TreeNode(textbook.TextBookID))
                Next
            Next
        Next

        trvTest.EndUpdate()
    End Sub
End Class
